File: src/commands/filter.ts
import { Command } from 'commander';
import { getContext, handleError } from './shared.js';
import { getFilter, setFilter, clearFilter } from '../db/queries.js';

export function filterCommand(): Command {
  const cmd = new Command('filter')
    .description('Manage message filter preferences');

  cmd.command('set')
    .description('Set filter preferences')
    .option('--mentions <pattern>', 'Mention pattern (e.g., "claude" or "claude,pm")')
    .action((options, cmd) => {
      try {
        const { db } = getContext(cmd);
        const agentId = process.env.MM_AGENT_ID;
        if (!agentId) {
          throw new Error('MM_AGENT_ID not set. Run mm new first.');
        }

        setFilter(db, {
          agent_id: agentId,
          mentions_pattern: options.mentions ?? null,
        });

        console.log('Filter updated');
        db.close();
      } catch (error) {
        handleError(error);
      }
    });

  cmd.command('show')
    .description('Show current filter')
    .action((options, cmd) => {
      try {
        const { db } = getContext(cmd);
        const agentId = process.env.MM_AGENT_ID;
        if (!agentId) {
          throw new Error('MM_AGENT_ID not set');
        }

        const filter = getFilter(db, agentId);
        if (!filter) {
          console.log('No filter set');
        } else {
          console.log('Current filter:');
          if (filter.mentions_pattern) console.log(`  Mentions: ${filter.mentions_pattern}`);
        }
        db.close();
      } catch (error) {
        handleError(error);
      }
    });

  cmd.command('clear')
    .description('Clear all filters')
    .action((options, cmd) => {
      try {
        const { db } = getContext(cmd);
        const agentId = process.env.MM_AGENT_ID;
        if (!agentId) {
          throw new Error('MM_AGENT_ID not set');
        }

        clearFilter(db, agentId);
        console.log('Filter cleared');
        db.close();
      } catch (error) {
        handleError(error);
      }
    });

  return cmd;
}

File: internal/command/filter.go
package command

import (
	"fmt"
	"os"

	"github.com/adamavenir/mini-msg/internal/db"
	"github.com/adamavenir/mini-msg/internal/types"
	"github.com/spf13/cobra"
)

// NewFilterCmd creates the filter command.
func NewFilterCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "filter",
		Short: "Manage message filter preferences",
	}

	cmd.AddCommand(newFilterSetCmd())
	cmd.AddCommand(newFilterShowCmd())
	cmd.AddCommand(newFilterClearCmd())

	return cmd
}

func newFilterSetCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "set",
		Short: "Set filter preferences",
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			agentID := os.Getenv("MM_AGENT_ID")
			if agentID == "" {
				return writeCommandError(cmd, fmt.Errorf("MM_AGENT_ID not set. Run mm new first."))
			}

			mentions, _ := cmd.Flags().GetString("mentions")
			var mentionsPtr *string
			if mentions != "" {
				mentionsPtr = &mentions
			}

			if err := db.SetFilter(ctx.DB, types.Filter{
				AgentID:         agentID,
				MentionsPattern: mentionsPtr,
			}); err != nil {
				return writeCommandError(cmd, err)
			}

			fmt.Fprintln(cmd.OutOrStdout(), "Filter updated")
			return nil
		},
	}

	cmd.Flags().String("mentions", "", `Mention pattern (e.g., "claude" or "claude,pm")`)
	return cmd
}

func newFilterShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show",
		Short: "Show current filter",
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			agentID := os.Getenv("MM_AGENT_ID")
			if agentID == "" {
				return writeCommandError(cmd, fmt.Errorf("MM_AGENT_ID not set"))
			}

			filter, err := db.GetFilter(ctx.DB, agentID)
			if err != nil {
				return writeCommandError(cmd, err)
			}

			if filter == nil {
				fmt.Fprintln(cmd.OutOrStdout(), "No filter set")
				return nil
			}

			fmt.Fprintln(cmd.OutOrStdout(), "Current filter:")
			if filter.MentionsPattern != nil {
				fmt.Fprintf(cmd.OutOrStdout(), "  Mentions: %s\n", *filter.MentionsPattern)
			}
			return nil
		},
	}
}

func newFilterClearCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "clear",
		Short: "Clear all filters",
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			agentID := os.Getenv("MM_AGENT_ID")
			if agentID == "" {
				return writeCommandError(cmd, fmt.Errorf("MM_AGENT_ID not set"))
			}

			if err := db.ClearFilter(ctx.DB, agentID); err != nil {
				return writeCommandError(cmd, err)
			}

			fmt.Fprintln(cmd.OutOrStdout(), "Filter cleared")
			return nil
		},
	}
}
